GEODE-734: gfsh export stack-traces should not require an output file with extension .txt

deepakddixit opened the pull request:
Removed mandatory txt file extension when exporting stack trace using gfsh command export stack-traces.
Added message warning user about possible overwrite of file if already present under locator directory.
Added test to check if non txt extension file is allowed.
All tests are passing

metatype commented:
Thanks for the contribution!  However, I don't think we should require user input in order to gather stack traces.  Quite often a user will automate collection (`gfsh export stack-traces ...`) and will not be able to provide input.

Here's a few ways we could extend the current behavior:
- Provide an option to auto-generate a unique name
- Provide an option that would cause the export to fail if the provided file name is already present.

kirklund commented on geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/ShowStackTraceDUnitTest.java:
```diff
@@ -90,6 +90,20 @@ public void run() {
     setupSystem();
 
+    // Test non txt extension file is allowed
+    File stacktracesFile = new File("allStackTraces.log");
```
Files should be created in a TemporaryFolder. Please see TemporaryFolder on https://github.com/junit-team/junit4/wiki/Rules.

We have a version of TemporaryFolder which works in DUnit (DistributedTests). See org.apache.geode.test.junit.rules.serializable.SerializableTemporaryFolder.

The TemporaryFolder Rule automatically takes care of deleting files and directories after each test.

deepakddixit replied:
Thanks @kirklund. Will use TemporaryFolder to work with files.

kirklund commented on geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/ShowStackTraceDUnitTest.java:
```diff
@@ -149,4 +163,92 @@ public void testExportStacktrace() throws ClassNotFoundException, IOException {
+   * @throws ClassNotFoundException
+   * @throws IOException
+   */
+  @Test
```
This looks like three different tests all inside one test. It's better to separate these so they're shorter, more readable and pass or fail independently of one another.

If you're worried about how long it takes to setup/teardown or run tests, then you really should consider testing this feature in a UnitTest or IntegrationTest.

UnitTest would use Mockito to help you isolate the one class for test (UnitTest shouldn't touch file system or network). IntegrationTest would allow you to touch file system or network, test group of classes, or create one Geode member (in one JVM). DistributedTest should only be used when you really need more than one member (multiple JVMs).

deepakddixit commented:
@metatype and @kirklund I have addressed the review comments. Kindly have a look.
precheckin is successful. I am not sure why travis is failed with spotlessJavaCheck, at my box it is successful. I tried spotlessApply but it shows no changes.

kirklund commented on geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/ShowStackTraceDUnitTest.java:
```diff
@@ -149,4 +143,126 @@ public void testExportStacktrace() throws ClassNotFoundException, IOException {
+      assertTrue(((String) exportCommandResult.getResultData().getGfJsonObject()
+          .getJSONObject("content").getJSONArray("message").get(0))
+              .contains("file " + stacktracesFile.getCanonicalPath() + " already present"));
+    } catch (GfJsonException e) {
```
I have one more change to request to the two tests with try-catch blocks. Please change the method signature to throws Exception and delete the try-catch block. JUnit handles this better than using fail with e.getCause() or e.getMessage(). With JUnit handling it, it's much easier to debug if these tests ever fail. Thanks!

kirklund commented:
I created feature/GEODE-734 branch and pulled in all changes. I also did a git rebase and combined all commits into a single commit:

commit 2c5a0a1688b821f95584ef4b6610c65d0433d2e7
Author: Deepak Dixit <deepakdixit2011@gmail.com>
Date:   Sat Nov 26 19:11:52 2016 +0530

    GEODE-734: Remove txt file restriction when exporting stack traces
    
    Adding throws clause to signature instead of catching it in Test.
    
    Separating test scenarios into multiple test cases. Creating files in
    TemporaryFolder.
    
    Applying spotless checks
    
    Updated export stack-traces in controller. Removed compulsory file
    option. If file is not provided then name is generated. Added one
    boolean option fail-if-file-present, if true and file already exists
    then command will fail.
    
    Updated Offline help file.
    
    Added test cases to verify new added options
    
    Removed compulsory file option. If file is not provided then name is
    generated. Added one boolean option fail-if-file-present, if true and
    file already exists then command will fail.
    
    Added message warning user about possible overwrite of file if already
    present.
    
    Added test to check if non txt extension file is allowed.
    
    This closes #297

kirklund commented:
I'm running precheckin again just to make sure. Precheckin should be done in about 3 hours and if that looks clean I'll merge this to develop.

deepakddixit commented:
Thank you @kirklund and @metatype for review and merge.
